File: ios/Sources/GhostCrypto/DoubleRatchet.swift
import CryptoKit
import Foundation
import Security

// MARK: - Key Pair Generator Protocol

public protocol KeyPairGenerating: AnyObject {
    func generateKeyPair() -> P256.KeyAgreement.PrivateKey
}

/// Default generator: random P-256 keys
public final class RandomKeyPairGenerator: KeyPairGenerating {
    public init() {}
    public func generateKeyPair() -> P256.KeyAgreement.PrivateKey {
        P256.KeyAgreement.PrivateKey()
    }
}

// MARK: - Encrypted Message Result

public struct EncryptedMessage {
    public let wireBase64: String
    #if DEBUG
    // Test-only: reveals the message key used to produce `wireBase64`.
    // Compiled out in release builds — keep out of any production code path.
    public let debugMessageKey: Data?
    #endif
}

// MARK: - Double Ratchet

public enum RatchetRole { case host, guest }

public struct DoubleRatchet {

    // Ratchet state
    private var dhs: P256.KeyAgreement.PrivateKey     // our DH keypair
    private var dhr: P256.KeyAgreement.PublicKey?      // their DH public key
    private var rk: SymmetricKey                       // root key
    private var cks: SymmetricKey?                     // sending chain key
    private var ckr: SymmetricKey?                     // receiving chain key
    private var ns: UInt32 = 0                         // send counter
    private var nr: UInt32 = 0                         // receive counter
    private var pn: UInt32 = 0                         // previous chain length

    // Skipped message keys: (dhPubRaw hex + ":" + messageNumber) → messageKey
    private var mkSkipped: [String: SymmetricKey] = [:]
    private let maxSkip: UInt32 = 100

    // Last seen DH key (for detecting ratchet triggers)
    private var lastDHr: Data?

    private let keyGen: KeyPairGenerating

    // MARK: - Init

    /// Initialize Double Ratchet.
    /// HOST: generates new DHs, does initial ratchet with guest's public key.
    /// GUEST: uses own ephemeral key as DHs, waits for first message.
    public init(
        role: RatchetRole,
        sharedKey: SymmetricKey,
        ourKeyPair: P256.KeyAgreement.PrivateKey,
        theirPublicKey: P256.KeyAgreement.PublicKey?,
        keyPairGenerator: KeyPairGenerating? = nil
    ) {
        self.keyGen = keyPairGenerator ?? RandomKeyPairGenerator()

        switch role {
        case .host:
            // HOST: DHs = provided keypair, DHr = guest's public key
            self.dhs = ourKeyPair
            self.dhr = theirPublicKey
            self.rk = sharedKey

            // Initial DH ratchet step — set up sending chain
            if let peerPub = theirPublicKey {
                let dhOutput = try! dhs.sharedSecretFromKeyAgreement(with: peerPub).rawData
                let result = CryptoUtils.rootKDF(rootKey: self.rk, dhOutput: dhOutput)
                self.rk = result.newRootKey
                self.cks = result.chainKey
            }
            self.ckr = nil
            self.lastDHr = theirPublicKey.map { dhPublicKeyRaw($0) }

        case .guest:
            // GUEST: DHs = own ephemeral, DHr = nil, RK = shared key
            self.dhs = ourKeyPair
            self.dhr = nil
            self.rk = sharedKey
            self.cks = nil
            self.ckr = nil
        }
    }

    // MARK: - Encrypt

    /// Encrypt plaintext and return wire-format base64 string.
    public mutating func encrypt(
        plaintext: String,
        deterministicNonce: Data? = nil,
        deterministicPadByte: UInt8? = nil
    ) throws -> EncryptedMessage {
        guard let cks = self.cks else {
            throw RatchetError.noSendingChain
        }

        // Chain KDF → message key
        let chain = CryptoUtils.chainKDF(chainKey: cks)
        self.cks = chain.nextChainKey

        // Pad message
        let padded: Data
        if let padByte = deterministicPadByte {
            padded = MessagePadding.pad(plaintext, deterministicPadByte: padByte)
        } else {
            padded = MessagePadding.pad(plaintext)
        }

        // Build header
        let dhPubRaw = dhPublicKeyRaw(dhs.publicKey)
        let header = WireFormat.buildHeader(dhPublicKeyRaw: dhPubRaw, pn: pn, n: ns)

        // Nonce: random or deterministic for tests
        let nonceData: Data
        if let dn = deterministicNonce {
            nonceData = dn
        } else {
            var randomNonce = Data(count: 12)
            randomNonce.withUnsafeMutableBytes { buf in
                _ = SecRandomCopyBytes(kSecRandomDefault, 12, buf.baseAddress!)
            }
            nonceData = randomNonce
        }

        let nonce = try AES.GCM.Nonce(data: nonceData)

        // AES-256-GCM encrypt with header as AAD
        let sealed = try AES.GCM.seal(padded, using: chain.messageKey, nonce: nonce, authenticating: header)

        // Build wire message
        let wire = WireFormat.buildMessage(
            header: header,
            nonce: nonceData,
            ciphertext: sealed.ciphertext,
            tag: Data(Array(sealed.tag))
        )

        ns += 1

        #if DEBUG
        let mkData = chain.messageKey.rawData
        return EncryptedMessage(
            wireBase64: wire.base64EncodedString(),
            debugMessageKey: mkData
        )
        #else
        return EncryptedMessage(
            wireBase64: wire.base64EncodedString()
        )
        #endif
    }

    // MARK: - Decrypt

    /// Decrypt a wire-format base64 string, returning the plaintext.
    public mutating func decrypt(wireBase64: String) throws -> String {
        guard let wireData = Data(base64Encoded: wireBase64) else {
            throw RatchetError.invalidBase64
        }

        let parsed = try WireFormat.parseMessage(wireData)
        let headerParsed = try WireFormat.parseHeader(parsed.header)

        // Reconstruct peer's public key from header (add 04 prefix for x963)
        let peerDHPubRaw = headerParsed.dhPublicKeyRaw
        let peerDHPubX963 = Data([0x04]) + peerDHPubRaw
        let peerPub = try P256.KeyAgreement.PublicKey(x963Representation: peerDHPubX963)

        // Check if this is a skipped message
        let skipKey = peerDHPubRaw.hexString + ":" + String(headerParsed.n)
        if let mk = mkSkipped[skipKey] {
            mkSkipped.removeValue(forKey: skipKey)
            return try decryptWithKey(mk, parsed: parsed)
        }

        // Check if we need a DH ratchet step
        let peerRawHex = peerDHPubRaw.hexString
        let currentDHrHex = lastDHr?.hexString

        if peerRawHex != currentDHrHex {
            // Skip missed messages from current receiving chain
            if self.ckr != nil {
                try skipMessages(until: headerParsed.pn, dhPubRaw: lastDHr!)
            }

            // DH ratchet step
            self.pn = self.ns
            self.ns = 0
            self.nr = 0
            self.dhr = peerPub
            self.lastDHr = peerDHPubRaw

            // Receiving chain
            let dhRecv = try dhs.sharedSecretFromKeyAgreement(with: peerPub).rawData
            let recvResult = CryptoUtils.rootKDF(rootKey: rk, dhOutput: dhRecv)
            self.rk = recvResult.newRootKey
            self.ckr = recvResult.chainKey

            // New sending keypair + chain
            self.dhs = keyGen.generateKeyPair()
            let dhSend = try dhs.sharedSecretFromKeyAgreement(with: peerPub).rawData
            let sendResult = CryptoUtils.rootKDF(rootKey: rk, dhOutput: dhSend)
            self.rk = sendResult.newRootKey
            self.cks = sendResult.chainKey
        }

        // Skip any missed messages in current receiving chain
        try skipMessages(until: headerParsed.n, dhPubRaw: peerDHPubRaw)

        // Chain KDF for this message
        guard let ckr = self.ckr else {
            throw RatchetError.noReceivingChain
        }
        let chain = CryptoUtils.chainKDF(chainKey: ckr)
        self.ckr = chain.nextChainKey
        self.nr += 1

        return try decryptWithKey(chain.messageKey, parsed: parsed)
    }

    // MARK: - Private Helpers

    private mutating func skipMessages(until n: UInt32, dhPubRaw: Data) throws {
        guard var ckr = self.ckr else { return }

        while nr < n {
            guard mkSkipped.count < maxSkip else {
                throw RatchetError.tooManySkippedMessages
            }
            let chain = CryptoUtils.chainKDF(chainKey: ckr)
            let key = dhPubRaw.hexString + ":" + String(nr)
            mkSkipped[key] = chain.messageKey
            ckr = chain.nextChainKey
            nr += 1
        }
        self.ckr = ckr
    }

    private func decryptWithKey(_ messageKey: SymmetricKey, parsed: WireFormat.ParsedMessage) throws -> String {
        let nonce = try AES.GCM.Nonce(data: parsed.nonce)
        let box = try AES.GCM.SealedBox(nonce: nonce, ciphertext: parsed.ciphertext, tag: parsed.tag)
        let padded = try AES.GCM.open(box, using: messageKey, authenticating: parsed.header)
        return MessagePadding.unpad(padded)
    }
}

// MARK: - Errors

public enum RatchetError: Error {
    case noSendingChain
    case noReceivingChain
    case invalidBase64
    case tooManySkippedMessages
    case replayDetected
}

// MARK: - Helpers

/// Extract 64-byte raw public key (no 04 prefix) from P256 public key
private func dhPublicKeyRaw(_ key: P256.KeyAgreement.PublicKey) -> Data {
    let x963 = key.x963Representation
    return Data(x963.dropFirst()) // Remove 04 prefix
}

File: android/crypto/src/main/kotlin/com/kordar/ghostchat/core/crypto/DoubleRatchet.kt
package com.kordar.ghostchat.core.crypto

import java.security.PrivateKey
import java.security.PublicKey
import java.security.SecureRandom
import java.util.Base64

enum class RatchetRole { HOST, GUEST }

interface KeyPairGenerator {
    fun generate(): CryptoUtils.ECKeyPair
}

class RandomKeyPairGenerator : KeyPairGenerator {
    override fun generate(): CryptoUtils.ECKeyPair = CryptoUtils.generateKeyPair()
}

class EncryptedMessage internal constructor(
    val wireBase64: String,
    @VisibleForTesting internal val debugMessageKey: ByteArray? = null
)

class DoubleRatchet(
    role: RatchetRole,
    sharedKey: ByteArray,
    ourKeyPair: CryptoUtils.ECKeyPair,
    theirPublicKey: PublicKey?,
    private val keyGen: KeyPairGenerator = RandomKeyPairGenerator()
) {
    private var dhs: CryptoUtils.ECKeyPair = ourKeyPair
    private var dhr: PublicKey? = theirPublicKey
    private var rk: ByteArray = sharedKey.copyOf()
    private var cks: ByteArray? = null
    private var ckr: ByteArray? = null
    private var ns: Int = 0
    private var nr: Int = 0
    private var pn: Int = 0

    private val mkSkipped = mutableMapOf<String, ByteArray>() // "dhPubRawHex:N" → messageKey
    private val maxSkip = 100
    private var lastDHr: ByteArray? = null

    /**
     * Re-hydrate a ratchet that was previously serialised via [exportedState].
     * Used to persist per-contact sessions in SQLCipher across app restarts.
     */
    constructor(
        stateBytes: ByteArray,
        keyGen: KeyPairGenerator = RandomKeyPairGenerator()
    ) : this(
        role = RatchetRole.HOST,       // placeholder — values overwritten below
        sharedKey = ByteArray(32),
        ourKeyPair = CryptoUtils.generateKeyPair(),
        theirPublicKey = null,
        keyGen = keyGen
    ) {
        val snap = DoubleRatchetState.deserialize(stateBytes)
        dhs = CryptoUtils.keyPairFromPrivateBytes(snap.dhsPrivateBytes)
        dhr = snap.dhrRaw?.let { CryptoUtils.publicKeyFromBytes(it) }
        rk = snap.rk.copyOf()
        cks = snap.cks?.copyOf()
        ckr = snap.ckr?.copyOf()
        ns = snap.ns; nr = snap.nr; pn = snap.pn
        lastDHr = snap.lastDHr?.copyOf()
        mkSkipped.clear()
        mkSkipped.putAll(snap.mkSkipped)
    }

    /** Serialised opaque ratchet state suitable for persistence in SQLCipher. */
    val exportedState: ByteArray
        get() = DoubleRatchetState.serialize(
            DoubleRatchetState.Snapshot(
                dhsPrivateBytes = dhs.privateKeyBytes,
                dhrRaw = lastDHr?.copyOf(), // current remote DH raw if any
                rk = rk.copyOf(),
                cks = cks?.copyOf(),
                ckr = ckr?.copyOf(),
                ns = ns, nr = nr, pn = pn,
                lastDHr = lastDHr?.copyOf(),
                mkSkipped = mkSkipped.mapValues { it.value.copyOf() }
            )
        )

    init {
        when (role) {
            RatchetRole.HOST -> {
                if (theirPublicKey != null) {
                    val dhOutput = CryptoUtils.ecdhSharedSecret(dhs.privateKey, theirPublicKey)
                    val result = CryptoUtils.rootKDF(rk, dhOutput)
                    rk = result.newRootKey
                    cks = result.chainKey
                    lastDHr = publicKeyRaw(theirPublicKey)
                }
            }
            RatchetRole.GUEST -> {
                // Wait for first message
            }
        }
    }

    fun encrypt(
        plaintext: String,
        deterministicNonce: ByteArray? = null,
        deterministicPadByte: Byte? = null
    ): EncryptedMessage {
        val currentCKs = cks ?: throw IllegalStateException("No sending chain")

        val chain = CryptoUtils.chainKDF(currentCKs)
        cks = chain.nextChainKey

        val padded = if (deterministicPadByte != null) {
            MessagePadding.pad(plaintext, deterministicPadByte)
        } else {
            MessagePadding.pad(plaintext)
        }

        val dhPubRaw = dhs.publicKeyRaw
        val header = WireFormat.buildHeader(dhPubRaw, pn, ns)

        val nonce = deterministicNonce ?: ByteArray(12).also { SecureRandom().nextBytes(it) }

        // AES-GCM: BouncyCastle returns ciphertext+tag concatenated
        val ctWithTag = CryptoUtils.aesGcmEncrypt(chain.messageKey, nonce, padded, header)
        val ct = ctWithTag.copyOfRange(0, ctWithTag.size - 16)
        val tag = ctWithTag.copyOfRange(ctWithTag.size - 16, ctWithTag.size)

        val wire = WireFormat.buildMessage(header, nonce, ct, tag)
        ns++

        return EncryptedMessage(
            wireBase64 = Base64.getEncoder().encodeToString(wire),
            debugMessageKey = chain.messageKey
        )
    }

    fun decrypt(wireBase64: String): String {
        val wireData = Base64.getDecoder().decode(wireBase64)
        val parsed = WireFormat.parseMessage(wireData)
        val headerParsed = WireFormat.parseHeader(parsed.header)

        val peerDHPubRaw = headerParsed.dhPublicKeyRaw
        val peerPub = CryptoUtils.publicKeyFromBytes(peerDHPubRaw)

        // Check skipped messages
        val skipKey = peerDHPubRaw.toHexString() + ":" + headerParsed.n
        mkSkipped.remove(skipKey)?.let { mk ->
            return decryptWithKey(mk, parsed)
        }

        // Check if ratchet needed
        val peerRawHex = peerDHPubRaw.toHexString()
        val currentDHrHex = lastDHr?.toHexString()

        if (peerRawHex != currentDHrHex) {
            // Skip missed messages
            if (ckr != null && lastDHr != null) {
                skipMessages(headerParsed.pn, lastDHr!!)
            }

            // DH ratchet
            pn = ns
            ns = 0
            nr = 0
            dhr = peerPub
            lastDHr = peerDHPubRaw.copyOf()

            // Receiving chain
            val dhRecv = CryptoUtils.ecdhSharedSecret(dhs.privateKey, peerPub)
            val recvResult = CryptoUtils.rootKDF(rk, dhRecv)
            rk = recvResult.newRootKey
            ckr = recvResult.chainKey

            // New sending keypair + chain
            dhs = keyGen.generate()
            val dhSend = CryptoUtils.ecdhSharedSecret(dhs.privateKey, peerPub)
            val sendResult = CryptoUtils.rootKDF(rk, dhSend)
            rk = sendResult.newRootKey
            cks = sendResult.chainKey
        }

        // Skip missed messages in current chain
        skipMessages(headerParsed.n, peerDHPubRaw)

        // Chain step
        val currentCKr = ckr ?: throw IllegalStateException("No receiving chain")
        val chain = CryptoUtils.chainKDF(currentCKr)
        ckr = chain.nextChainKey
        nr++

        return decryptWithKey(chain.messageKey, parsed)
    }

    private fun skipMessages(until: Int, dhPubRaw: ByteArray) {
        var currentCKr = ckr ?: return

        while (nr < until) {
            check(mkSkipped.size < maxSkip) { "Too many skipped messages" }
            val chain = CryptoUtils.chainKDF(currentCKr)
            val key = dhPubRaw.toHexString() + ":" + nr
            mkSkipped[key] = chain.messageKey
            currentCKr = chain.nextChainKey
            nr++
        }
        ckr = currentCKr
    }

    private fun decryptWithKey(messageKey: ByteArray, parsed: WireFormat.ParsedMessage): String {
        val ctWithTag = parsed.ciphertext + parsed.tag
        val padded = CryptoUtils.aesGcmDecrypt(messageKey, parsed.nonce, ctWithTag, parsed.header)
        return MessagePadding.unpad(padded)
    }

    private fun publicKeyRaw(key: PublicKey): ByteArray {
        val full = (key as org.bouncycastle.jcajce.provider.asymmetric.ec.BCECPublicKey)
            .q.getEncoded(false)
        return full.copyOfRange(1, 65)
    }
}
